scheduler/scheduler.go

```diff
@@ -107,6 +117,111 @@ type managesWork interface {
 	Work(job) queuedJob
 }
 
+// Implemented as a separate function so that defer calls
+// are properly handled and cleanup done properly
+// (like for the removal of temporary Yaml to JSON conversion)
+func autoDiscoverTasks(taskFiles []os.FileInfo, fullPath string,
+	fp func(sch schedule.Schedule,
+		wfMap *wmap.WorkflowMap,
+		startOnCreate bool,
+		opts ...core.TaskOption) (core.Task, core.TaskErrors)) {
+	// Note that the list of files is sorted by name due to ioutil.ReadDir
+	// default behaviour. See go doc ioutil.ReadDir
+	for _, file := range taskFiles {
+		f, err := os.Open(path.Join(fullPath, file.Name()))
+		if err != nil {
+			log.WithFields(log.Fields{
+				"_block":           "autoDiscoverTasks",
+				"_module":          "scheduler",
+				"autodiscoverpath": fullPath,
+				"task":             file.Name(),
+			}).Error("Opening file ", err)
+			continue
+		}
+		defer f.Close()
+		if !strings.HasSuffix(file.Name(), ".json") {
+			fc, err := ioutil.ReadAll(f)
+			if err != nil {
+				log.WithFields(
+					log.Fields{
+						"_block":           "autoDiscoverTasks",
+						"_module":          "scheduler",
+						"autodiscoverpath": fullPath,
+						"task":             file.Name(),
+					}).Error("Reading Yaml file ", err)
+				continue
+			}
+			js, err := yaml.YAMLToJSON(fc)
+			if err != nil {
+				log.WithFields(
+					log.Fields{
+						"_block":           "autoDiscoverTasks",
+						"_module":          "scheduler",
+						"autodiscoverpath": fullPath,
+						"task":             file.Name(),
+					}).Error("Parsing Yaml file ", err)
+				continue
+			}
+			tfile, err := ioutil.TempFile(os.TempDir(), "yaml2json")
```

Comment:
@obourdon I was planning to look at this some more today and I will take another look at this section to wrap my head around it. Thanks for pointing out what I was missing.